File: ts/game/queue.ts
import { Mino } from "./data.js";
import { RNG } from "./rng.js";

export enum Bag {
  Bag7 = "7-bag",
}

export class BagData {
  static getCycle(bag: Bag): Mino[] {
    switch (bag) {
      case Bag.Bag7:
        return [Mino.Z, Mino.L, Mino.O, Mino.S, Mino.I, Mino.J, Mino.T];
    }
  }
}

export class Queue {
  public bag: Bag;
  public rng: RNG;
  public minSize: number;
  public queue: Mino[];

  constructor(bag: Bag, seed: number, minSize: number, initial: Mino[]) {
    console.assert(minSize >= 16, "Bag min size must be at least 16");
    this.bag = bag;
    this.rng = new RNG(seed);
    this.minSize = minSize;
    this.queue = [...initial];

    while (this.queue.length < minSize) {
      const cycle = this.rng.shuffle(BagData.getCycle(bag));
      this.queue.push(...cycle);
    }
  }

  shift(): Mino {
    const res = this.queue.shift() || Mino.I;

    while (this.queue.length < this.minSize) {
      const cycle = this.rng.shuffle(BagData.getCycle(this.bag));
      this.queue.push(...cycle);
    }

    return res;
  }

  getFront16(): Mino[] {
    const res: Mino[] = new Array(16);

    for (let i = 0; i < 16; i++) {
      res[i] = this.queue[i] || Mino.I;
    }

    return res;
  }

  clone(): Queue {
    const cloned = new Queue(this.bag, 0, this.minSize, []);
    cloned.rng = new RNG(this.rng.seed);
    cloned.rng.index = this.rng.index;
    cloned.queue = [...this.queue];
    return cloned;
  }
}

File: engine/src/game/queue.rs
use std::collections::VecDeque;

use serde::Deserialize;
use triangle::engine::queue::Mino;

use super::rng::RNG;

#[derive(Deserialize, Clone, Copy, Debug)]
pub enum Bag {
  #[serde(rename = "7-bag")]
  Bag7,
}

impl Bag {
  pub fn get_cycle(&self) -> Vec<Mino> {
    match self {
      Bag::Bag7 => Vec::from([
        Mino::Z,
        Mino::L,
        Mino::O,
        Mino::S,
        Mino::I,
        Mino::J,
        Mino::T,
      ]),
    }
  }
}

#[derive(Clone)]
pub struct Queue<const N: usize> {
  pub bag: Bag,
  pub rng: RNG,
  pub queue: VecDeque<Mino>,
}

impl<const N: usize> Queue<N> {
  pub fn new(bag: Bag, seed: u64, initial: Vec<Mino>) -> Self {
    let mut rng = RNG::new(seed);

    let mut queue: VecDeque<Mino> = VecDeque::with_capacity(N + 7);

    for m in initial.iter() {
      queue.push_back(*m);
    }

    while queue.len() < N {
      for mino in rng.shuffle(bag.get_cycle()) {
        queue.push_back(mino);
      }
    }

    Queue { bag, rng, queue }
  }

  pub fn shift(&mut self) -> Mino {
    let res = self
      .queue
      .pop_front()
      .unwrap_or_else(|| unreachable!("Queue is empty!"));

    while self.queue.len() < N {
      for mino in self.rng.shuffle(self.bag.get_cycle()) {
        self.queue.push_back(mino);
      }
    }

    res
  }

  pub fn as_array(&self) -> [Mino; N] {
    std::array::from_fn(|i| *self.queue.get(i).unwrap_or(&Mino::I))
  }
}
